src/input-states.js

```diff
@@ -0,0 +1,242 @@
+    var port = node.inPorts[portName];
+    if (!isMultiPort(port)) throw Error("This port is not addressable/multi: " + portName);
+    else return port.listAttached().map(function(socketId){
+        return getPortStateBySocketId(node, vnid, portName, socketId);
+    });
+}
+
+/**
+ * Returns the state of this port for the given socketId. This can be used to
+ * look up the last state when checking to see if the state has changed.
+ * Note that the socketId is not neccessarily the same as the array index
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the name of the addressable/multi port for this node
+ * @param socketId the socketId provided to a port data event
+ */
+function getPortStateBySocketId(node, vnid, portName, socketId) {
+    if (socketId == null)
+        return getPortState(node, vnid, portName);
+    else if (!isMultiPort(node.inPorts[portName]))
+        throw Error("This port is not addressable/multi: " + portName);
+    else if (ensureVniStateExists(node, vnid, portName) &&
+            node.vnis[vnid].inputStates[portName][socketId])
+        return node.vnis[vnid].inputStates[portName][socketId];
+    // By default states that come in on vnid '' apply to all VNIs
+    else if (vnid) return getPortStateBySocketId(node, '', portName, socketId);
+    else return undefined;
+}
+
+/**
+ * Changes the state of each given port. If a port name is missing it is unmodified.
+ * This can be run in a test environment to reset the state of the ports.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portStates hash of portName to state
+ */
+function setAllPortStates(node, vnid, portStates) {
+    _.each(portStates, function(state, portName) {
+        setPortState(node, vnid, portName, state);
+    });
+    return node;
+}
+
+/**
+ * Changes the state of a port. When a new state is received on a port this
+ * should be called.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName name of the port on this node
+ * @param state the new state (or array of states if the port is multi/addressable) to assign to this port
+ */
+function setPortState(node, vnid, portName, state) {
+    var port = node.inPorts[portName];
+    if (_.isUndefined(state)) {
+        if (ensureVniStateExists(node, vnid, portName)) {
+            delete node.vnis[vnid].inputStates[portName];
+            if (_.isEmpty(node.vnis[vnid].inputStates)) {
+                delete node.vnis[vnid].inputStates;
+                if (_.isEmpty(node.vnis[vnid])) {
+                    delete node.vnis[vnid];
+                }
+            }
+        }
+    } else if (port.isMulti()) {
+        return setPortStateArray(node, vnid, portName, state);
+    } else {
+        if (!ensureVniStateExists(node, vnid, portName)) {
+            node.vnis = node.vnis || {};
```

Comment:
This should call a common ensureVniExists(node, vnid) function instead of creating he object directly, so that all VNI creation is done in one place.